File: csharp/tests/Int2Dds.Tests/Helpers/HelloWorldType.cs
using System;
using Int2Dds.Cdr;
using Int2Dds.Types;

namespace Int2Dds.Tests.Helpers
{
    /// <summary>
    /// IDL struct: HelloWorld { unsigned long index; string message; };
    /// Extensibility: FINAL, non-keyed.
    /// </summary>
    [DdsType("HelloWorld", 0, false)]
    public class HelloWorld : IDdsType
    {
        public uint Index { get; set; }
        public string Message { get; set; } = "";

        public HelloWorld() { }

        public HelloWorld(uint index, string message)
        {
            Index = index;
            Message = message;
        }

        public byte[] SerializeCdr()
        {
            return SerializeCdr(false);
        }

        public byte[] SerializeCdr(bool xcdr2)
        {
            var w = new CdrWriter(Extensibility.Final);
            SerializeCdr(w);
            return w.ToBytes();
        }

        public void SerializeCdr(CdrWriter w)
        {
            w.WriteU32(Index);
            w.WriteString(Message);
        }

        public static HelloWorld DeserializeCdr(byte[] data)
        {
            var r = new CdrReader(data);
            var index = r.ReadU32();
            var message = r.ReadString();
            return new HelloWorld(index, message);
        }

        private static readonly byte[] s_emptyKey = new byte[0];
        public byte[] SerializeKey() => s_emptyKey;
    }
}

File: python/tests/cross_process/hello_world_type.py
"""
HelloWorld data type for cross-process communication tests.

Matches the Rust HelloWorldType CDR layout:
    struct HelloWorld {
        unsigned long index;   // u32
        string message;
    };
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import ClassVar

from int2dds.cdr import CdrReader, CdrWriter, Extensibility


@dataclass
class HelloWorld:
    """IDL struct: HelloWorld (Extensibility: FINAL)"""

    _dds_type_name: ClassVar[str] = "HelloWorld"
    _extensibility: ClassVar[Extensibility] = Extensibility.FINAL
    _has_key: ClassVar[bool] = False

    index: int = 0
    message: str = ""

    def _serialize_cdr(self, xcdr2: bool = False) -> bytes:
        w = CdrWriter(extensibility=self._extensibility, xcdr2=xcdr2)
        w.write_u32(self.index)
        w.write_string(self.message)
        return w.to_bytes()

    @classmethod
    def _deserialize_cdr(cls, data: bytes) -> "HelloWorld":
        r = CdrReader(data)
        return cls(index=r.read_u32(), message=r.read_string())
